ARCHITECTURE.md

```diff
@@ -0,0 +1,196 @@
+
+- `${library}/*.h` are the public headers for the library. The same directory
+  contains the `${library}/*.cc` implementation files and `*_test.cc` for unit
+  tests. More often than not all the unit tests for `foo.{h,cc}` are found in
+  `foo_test.cc`, but there may be exceptions when the tests grow too large to
+  fit in a single file (we have lots of tests!).
+- `mocks/` also contains public APIs, typically this is a header-only library
+  defining mocks so application developers can mock the library in their own
+  tests. Note that some libraries do not have mocks.
+- `quickstart/` contains a small program showing how to use the library in a
+  larger project. This is intended for application developers wanting to, well,
+  use the libraries quickly.
+- `samples/` (sometimes `examples/`) contain code snippets that are extracted
+  both into the Doxygen reference documentation and to the cloud.google.com
+  site, as part of the service documentation. Some libraries have a single
+  program in this directory, others have smaller programs (or more samples).
+
+The remaining directories are implementation details, or at least only intended
+to be interesting for `google-cloud-cpp` developers and contributors.
+
+- `internal/` contains implementation details for the library. The files in
+  these directories are *not* intended for public consumption, but contributors
+  and developers need to change them, of course. The relationship between `.h`,
+  `.cc` and `_test.cc` files mirrors the one found in `${library}/`.
+- `benchmarks/` contains "macro benchmarks" for the library, typically programs
+  that interact with the production environment to verify the library meets the
+  latency, throughput, and CPU overhead (as in "low CPU overhead") requirements.
+- `ci/` may contain helper scripts used in the CI builds, for example, to
+  startup and shutdown any emulators used by the library.
+- `doc/` contains the landing page for any Doxygen documentation, as well as
+  additional Doxygen docs that are not extracted from the code comments.
+- `integration_tests/` (or simply `tests/` in some libraries) contain any
+  integration tests where we test the library using emulators and/or the
+  production environment.
+- `testing/` contains testing utilities shared by the unit tests and the
+  integration tests of the library. They are not intended to be part of the
+
+## The `*Client` classes
+
+The main interface for each library is the `${library}::*Client` class. Roughly
+speaking each "client" corresponds to a `service` in the `.proto` definitions
+for the service. It is common for services to have separate `service`
+definitions for "admin" operations, such as creating a new instance of the
+service, or changing permissions, as opposed to "normal" operations, such as
+inserting a new row, or publishing a new message. When "admin" services are
+defined, there are separate `${library}::Client` objects.
```

Comment:
Did you mean the previous name to be `${library}::Client` and this one the `${library}::*Client` glob?

Reply:
Fixed, thanks.